File: 0701. 二叉搜索树中的插入操作.java
// 1.尾递归
public TreeNode insertIntoBST(TreeNode root, int val) {
    // 找到插入位置
    if (root == null) {
        return new TreeNode(val);
    }
    if (val < root.val) {
        root.left = insertIntoBST(root.left, val);
    }
    else {
        root.right = insertIntoBST(root.right, val);
    }
    return root;
}

// 2.迭代
public TreeNode insertIntoBST(TreeNode root, int val) {
    TreeNode node = root;
    while (node != null) {
        if (val < node.val) {
            // 找到插入位置
            if (node.left == null) {
                node.left = new TreeNode(val);
                break;
            }
            node = node.left;
        }
        else {
            // 找到插入位置
            if (node.right == null) {
                node.right = new TreeNode(val);
                break;
            }
            node = node.right;
        }
    }
    return root;
}

File: 0701. 二叉搜索树中的插入操作.c
// 1.尾递归
struct TreeNode* insertIntoBST(struct TreeNode* root, int val) {
    // 找到插入位置
    if (!root) {
        struct TreeNode* node = malloc(sizeof(struct TreeNode));
        node->val = val;
        node->left = NULL;
        node->right = NULL;
        return node;
    }
    if (val < root->val) {
        root->left = insertIntoBST(root->left, val);
    }
    else {
        root->right = insertIntoBST(root->right, val);
    }        
    return root;
}

// 2.迭代
struct TreeNode* insertIntoBST(struct TreeNode* root, int val) {
    struct TreeNode* t = root;
    while (t) {
        if (val < t->val) {
            // 找到插入位置
            if (!t->left) {
                struct TreeNode* node = malloc(sizeof(struct TreeNode));
                node->val = val;
                node->left = NULL;
                node->right = NULL;
                t->left = node;
                break;
            }
            t = t->left;
        }
        else {
            // 找到插入位置
            if (!t->right) {
                struct TreeNode* node = malloc(sizeof(struct TreeNode));
                node->val = val;
                node->left = NULL;
                node->right = NULL;
                t->right = node;
                break;
            }
            t = t->right;
        }   
    }    
    return root;
}